### PR TITLE
fix port merge of kubernetes services in sidecar

### PR DESCRIPTION
**Pull request by ramaraochavali**
K8s servie merge in sidecar is broken since https://github.com/istio/istio/pull/48896. When a sidecar is defined with two egress listeners  pointing to two ports, the port merge does not happen

```
  egress:
  - hosts:
    - '*/foo.svc.cluster.local'
    port:
      name: tcp-port1
      number: 7443
      protocol: TCP
  - hosts:
    - '*/foo.svc.cluster.local'
    port:
      name: tcp-port2
      number: 7442
      protocol: TCP
```

With the above sidecar, the merged service will only have one port and hence only cluster is created.

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/model/sidecar_test.go [hunk 1]
```diff
@@ -2060,56 +2076,6 @@ func TestCreateSidecarScope(t *testing.T) {
 				},
 			},
 		},
-		{
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar_test.go:2063]: This test not relevant. Kubernetes service in the same namespace should be merged. The existing changed test verifies this behaviour

File: pilot/pkg/model/sidecar_test.go [hunk 2]
```diff
@@ -2201,7 +2167,7 @@ func TestCreateSidecarScope(t *testing.T) {
 			expectedServices: []*Service{
 				{
 					Hostname: "proxy",
-					Ports:    port7443,
+					Ports:    port744x,
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar_test.go:2170]: Again here, k8s service ports should be merged while taking precedence

**Comment by ramaraochavali** [comment]
This broke our existing deployments after upgrade.

**Comment by hzxuzhonghu** [comment]
can you describe how you merge svces in the egress listeners. As we can see in xds generator, sometimes we use services in egress listenrs, some times we use services in sidecarScope

**Comment by ramaraochavali** [comment]
See this example 

```
  egress:
  - hosts:
    - '*/foo.svc.cluster.local'
    port:
      name: tcp-port1
      number: 7443
      protocol: TCP
  - hosts:
    - '*/foo.svc.cluster.local'
    port:
      name: tcp-port2
      number: 7442
      protocol: TCP
  ```
  
It generates two listeners but only one cluster with outbound|7443||foo.svc.cluster.local. It does not generate cluster for 7442. This is broken in current release

File: pilot/pkg/model/sidecar.go [hunk 3]
```diff
@@ -929,19 +929,20 @@ func (sc *SidecarScope) appendSidecarServices(servicesAdded map[host.Name]sideca
 		sc.servicesByHostname[s.Hostname] = s
 	} else {
 		existing := foundSvc.svc
-		// We donot merge k8s service with any other services from other registries
-		if existing.Attributes.ServiceRegistry == provider.Kubernetes {
-			log.Debugf("Service %s/%s from registry %s ignored by %s/%s/%s", s.Attributes.Namespace, s.Hostname, s.Attributes.ServiceRegistry,
+		// We do not merge k8s service with any other services from other registries
+		if existing.Attributes.ServiceRegistry == provider.Kubernetes && s.Attributes.ServiceRegistry != provider.Kubernetes {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go:933]: how could this occur two services in k8s have same fqdn

File: pilot/pkg/model/sidecar.go [hunk 4]
```diff
@@ -929,19 +929,20 @@ func (sc *SidecarScope) appendSidecarServices(servicesAdded map[host.Name]sideca
...
 		// In some scenarios, there may be multiple Services defined for the same hostname due to ServiceEntry allowing
 		// arbitrary hostnames. In these cases, we want to pick the first Service, which is the oldest. This ensures
-		// newly created Services cannot take ownership unexpectedly.
-		// However, the Service is from Kubernetes it should take precedence over ones not. This prevents someone from
-		// "domain squatting" on the hostname before a Kubernetes Service is created.
-		if s.Attributes.ServiceRegistry == provider.Kubernetes {
-			log.Debugf("Service %s/%s from registry %s ignored by %s/%s/%s", existing.Attributes.Namespace, existing.Hostname, existing.Attributes.ServiceRegistry,
+		// newly created Services cannot take ownership unexpectedly. However, the Service is from Kubernetes it should
+		// take precedence over ones not. This prevents someone from "domain squatting" on the hostname before a Kubernetes Service is created.
+		if existing.Attributes.ServiceRegistry != provider.Kubernetes && s.Attributes.ServiceRegistry == provider.Kubernetes {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/sidecar.go:943]: The above comment apply here

File: pilot/pkg/model/sidecar.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:933]: Please try with this sidecar https://github.com/istio/istio/pull/50840#issuecomment-2095169376

Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:933]: I see how this occur, it is the egress listener with generate intermeditae svc with only one port from the original k8s svc.

**Review by hzxuzhonghu — COMMENTED** [review]
Can you add multiple sidecar egress listeners svc match test

File: pilot/pkg/model/sidecar.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:933]: Yes and we need to merge at that end

**Comment by ramaraochavali** [comment]
> Can you add multiple sidecar egress listeners svc match test

We already have that test https://github.com/istio/istio/blob/044be69f174a7afe8a827ddc4e9717df7ee97cd3/pilot/pkg/model/sidecar_test.go#L2214 - it did not fail because we do not specify service registry attribute. I fixed it

**Comment by hzxuzhonghu** [comment]
@ramaraochavali The test has service resides in different namespaces

**Comment by ramaraochavali** [comment]
I do not think that really matters much - even that failed without this. Any way, I have added additional test

**Pull request merged**